### ISSUE TITLE
IterDictIndexer constructor cannot take the meta field

### ISSUE THREAD

**Issue by Xiao0728** [terrier-org/pyterrier#250]
Hi,
I have tried to build the index using the following code, but the indexer.index() cannot take the meta as the pyterrier document examples shown:
**Examples using IterDictIndexer** from:https://pyterrier.readthedocs.io/en/latest/terrier-indexing.html

An iterdict can just be a list of dictionaries:

``docs = [ { 'docno' : 'doc1', 'text' : 'a b c' }  ]``
``iter_indexer = pt.IterDictIndexer("./index")``
``indexref1 = iter_indexer.index(docs, meta=['docno', 'text'], meta_lengths=[20, 4096])``

The code and error is:

![image](https://user-images.githubusercontent.com/43675140/145727738-4ae3f296-015e-4622-8322-b3b4e858bc2f.png)

**Comment by seanmacavaney** [comment]
Thanks for reporting! Passing other arguments into `index` (like meta, meta_lengths, etc.) is an anti-pattern that we've been meaning to fix -- we plan to move them to the IterDictIndexer constructor instead.

But in the meantime...

The problem arises because the ComposedPipeline (which is made when you use the `>>` operator) isn't expecting the `meta` and `meta_lengths` arguments to be provided. In the meantime, you should be able to work around this issue with something like:

```python
passage_transformer = pt.text.sliding(text_attr='text', prepend_attr=None)
def doc_iter():
    for batch in more_itertools.chunked(dataset.get_corpus_iter(), 100):
        for record in passage_transformer(pd.DataFrame(batch)).itertuples(index=False):
            yield record._asdict()
iter_index.erindex(doc_iter(), meta=['docno', 'text'], meta_lengths=[20, 4096])
```

(note: I didn't test this code)

**Comment by Xiao0728** [comment]
Thanks! 
"we plan to move them to the IterDictIndexer constructor instead" --Haven't moved yet, right? 
As I have also tested to use meta and meta length inside the IterDictIndexer constructor, but no actual field information is indexed.

**Comment by seanmacavaney** [comment]
> Haven't moved yet, right

Correct, we have not done this yet.

> As I have also tested to use meta and meta length inside the IterDictIndexer constructor, but no actual field information is indexed.

Can you elaborate and provide a test case? Here's a simple case showing the index and meta index working as expected:

```python
>>> import pyterrier as pt
>>> pt.init()
>>> indexer = pt.IterDictIndexer('./test_index')
>>> docs = [{'docno': '1', 'text': 'some text', 'some_meta': 'meta'}]
>>> index_ref = indexer.index(docs, meta=['docno', 'text', 'some_meta'], meta_lengths=[10, 100, 100])
>>> index = pt.IndexFactory.of(index_ref)
>>> index.metaIndex.getItem("docno", 0)
'1'
>>> index.metaIndex.getItem("text", 0)
'some text'
>>> index.metaIndex.getItem("some_meta", 0)
'meta'
>>> index.lexicon['text'].getFrequency()
1
```

**Comment by cmacdonald** [comment]
Sean, 

>>inside the IterDictIndexer constructor

>Here's a simple case showing the index and meta index working as expected

Your example has the metadata in the index() method - if we use compose, we indeed cant place it in the constructor of IterDictIndexer.

**Comment by seanmacavaney** [comment]
Yes, I know, but I gave a workaround that doesn't involve using compose -- see the code above.

**Comment by cmacdonald** [comment]
PR made:
https://github.com/terrier-org/pyterrier/pull/251/files

**Comment by cmacdonald** [comment]
Will be included in 0.7.2 release
